execution_engine/src/core/engine_state/execution_effect.rs

```diff
@@ -15,25 +17,27 @@ impl ExecutionEffect {
     }
 }
 
-impl From<&ExecutionEffect> for casper_types::ExecutionEffect {
-    fn from(effect: &ExecutionEffect) -> Self {
-        casper_types::ExecutionEffect {
-            operations: effect
-                .ops
-                .iter()
-                .map(|(key, op)| casper_types::Operation {
-                    key: key.to_formatted_string(),
-                    kind: op.into(),
-                })
-                .collect(),
-            transforms: effect
-                .transforms
-                .iter()
-                .map(|(key, transform)| casper_types::TransformEntry {
-                    key: key.to_formatted_string(),
-                    transform: transform.into(),
-                })
-                .collect(),
+impl From<ExecutionJournal> for ExecutionEffect {
+    fn from(journal: ExecutionJournal) -> Self {
+        let mut ops = AdditiveMap::new();
+        let mut transforms = AdditiveMap::new();
+        let journal: Vec<(Key, Transform)> = journal.into();
+        for (key, transform) in journal.into_iter() {
+            let op = match transform {
+                Transform::Failure(_) => continue,
+                Transform::Identity => Op::Read,
+                Transform::Write(_) => Op::Write,
+                Transform::AddInt32(_)
+                | Transform::AddUInt64(_)
+                | Transform::AddUInt128(_)
+                | Transform::AddUInt256(_)
+                | Transform::AddUInt512(_)
+                | Transform::AddKeys(_) => Op::Add,
+            };
+            ops.insert_add(key, op);
+            transforms.insert_add(key, transform);
```

Comment:
I don't think this will preserve insertion order. Let's change the backing store inside AdditiveMap into https://crates.io/crates/indexmap and add tests to ensure transforms are now in insertion order. Collapsing a Vec into AdditiveMap (aka HashMap) will break the insertion order.

Reply:
(As long as we are talking about types in execution_engine, as it isn't true for the casper_types ExecutionEffect), ExecutionEffect isn't used to preserve order. In those cases where order preservation is desirable, an ExecutionJournal is to be used.